src/morphology.rs

```diff
@@ -349,6 +349,184 @@ pub fn close_mut(image: &mut GrayImage, norm: Norm, k: u8) {
     erode_mut(image, norm, k);
 }
 
+/// A struct representing a mask used in morphological operations
+///
+/// the mask is represented by a list of the positions of its pixels
+/// relative to its center, with a maximum distance of 255
+/// along each axis.
+/// This means that in the most extreme case, the mask could have
+/// a size of 513 by 513 pixels.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Mask {
+    elements: Vec<(i16, i16)>,
+}
+
+impl Mask {
+    /// todo!()
+    pub fn square(radius: u8) -> Self {
+        Self {
+            elements: (-(radius as i16)..=(radius as i16))
+                .flat_map(|x| (-(radius as i16)..=(radius as i16)).map(move |y| (x, y)))
+                .collect(),
+        }
+    }
+
+    /// todo!()
+    pub fn diamond(radius: u8) -> Self {
+        Self {
+            elements: (-(radius as i16)..=(radius as i16))
+                .flat_map(|x| {
+                    ((x.abs() - radius as i16)..=(radius as i16 - x.abs())).map(move |y| (x, y))
+                })
+                .collect(),
+        }
+    }
+
+    /// todo!()
+    pub fn disk(radius: u8) -> Self {
+        let mut image = GrayImage::new(2 * radius as u32 + 1, 2 * radius as u32 + 1);
+        image.iter_mut().for_each(|p| *p = 0); // might not unnecessary
+        image.put_pixel(radius as u32, radius as u32, Luma::from([u8::MAX]));
+        dilate_mut(&mut image, Norm::L2, radius);
+        let im_ref = &image;
+        Self {
+            elements: (-(radius as i16)..=(radius as i16))
+                .flat_map(|x| {
+                    (-(radius as i16)..=(radius as i16))
+                        .filter(move |&y| im_ref.get_pixel(x as u32, y as u32).0[0] != 0)
+                        .map(move |y| (x, y))
+                })
+                .collect(),
+        }
+    }
+
+    /// todo!()
+    pub fn from_image(image: &GrayImage, center_x: u8, center_y: u8) -> Self {
+        assert!(
+            (image.width() - center_x as u32) < (u8::MAX as u32),
+            "all pixels of the mask must be at most {} pixels from the center",
+            u8::MAX
+        );
+        assert!(
+            (image.height() - center_y as u32) < (u8::MAX as u32),
+            "all pixels of the mask must be at most {} pixels from the center",
+            u8::MAX
+        );
+        Self {
+            elements: (0..image.width() as i16)
+                .flat_map(|x| {
+                    (0..(image.height() as i16))
+                        .filter(move |&y| image.get_pixel(x as u32, y as u32).0[0] != 0)
```

Comment:
Nearly all (or maybe all?) of your 'move' keywords look unnecessary as all the types you're using are `copy`.

Reply:
they are `copy`, which is why i can multiple `move` in the first place, but the compiler is unable to check whether the closure will outlive the 'x', therefore it needs to `move` a copy of the 'x' inside the closure. because the types are copy, this is practically free. (if we didn't move the 'x', the closure would have to carry a reference to it)

if you know of a way to remove them and make the borrow-checker happy, i'm willing to change things, but simply removing them is not an option.

Reply:
Can you give me an example of a compiler error when you remove one of the ‘move’s. I’m not understanding something here.

Reply:
```
 error[E0373]: closure may outlive the current function, but it borrows `x`, which is owned by the current function
   --> src\morphology.rs:396:30
    |
396 |                         .map(|y| (x, y))
    |                              ^^^  - `x` is borrowed here
    |                              |
    |                              may outlive borrowed value `x`
    |
note: closure is returned here
   --> src\morphology.rs:391:21
    |
391 | /                     (-(radius as i16)..=(radius as i16))
392 | |                         .filter(move |&y| {
393 | |                             (x.unsigned_abs() as u32).pow(2) + (y.unsigned_abs() as u32).pow(2)
394 | |                                 <= (radius as u32).pow(2)
395 | |                         })
396 | |                         .map(|y| (x, y))
    | |________________________________________^
help: to force the closure to take ownership of `x` (and any other referenced variables), use the `move` keyword
    |
396 |                         .map(move |y| (x, y))
    |                              ++++
```

Reply:
Ah yes, sorry - I'm talking nonsense!

This possibly reads slightly more clearly using the `cartesian_product` extension method from itertools:

```
       let range = -(radius as i16)..=(radius as i16);
        let elements: Vec<(i16, i16)> = range
            .clone()
            .cartesian_product(range)
            .filter(|(x, y)| {
                (x.unsigned_abs() as u32).pow(2) + (y.unsigned_abs() as u32).pow(2)
                    <= (radius as u32).pow(2)
            })
            .collect();
```

Pretty subjective though.

Reply:
thanks a ton ! if i knew that was an option, i would have used it, seems much clearer to me

Reply:
i changed it, but after thinking about it i think it could be that this change slows downs the functions a bit because the use of itertools prevents the compiler from figuring out the output vector size automatically. not really a speed critical place especially compared to the application of the masks, but i will try to test it later